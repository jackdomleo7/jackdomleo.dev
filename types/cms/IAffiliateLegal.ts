import { TRichText } from "./TRichText";

export interface IAffiliateLegal {
  legal_text: TRichText
}