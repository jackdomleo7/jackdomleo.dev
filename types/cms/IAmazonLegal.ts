import { TRichText } from "./TRichText";

export interface IAmazonLegal {
  legal_text: TRichText
}