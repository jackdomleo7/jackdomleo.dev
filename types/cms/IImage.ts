export interface IImage {
  url: string | null;
  alt: string | null;
  copyright: string | null;
  dimensions: {
    height: number;
    width: number;
  }
}