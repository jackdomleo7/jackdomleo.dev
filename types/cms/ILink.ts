export interface ILink {
  link_type: string;
  url?: string;
}