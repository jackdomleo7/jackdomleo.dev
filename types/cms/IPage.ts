export type TDocumentType = 'home' | 'projects' | 'blog' | 'about' | 'product' | 'amazon_affiliate_legal'

export interface IPage<DataType extends Record<string, any>, DocumentType extends TDocumentType> {
  alternate_languages: any[];
  first_publication_date: Date;
  last_publication_date: Date;
  href: string;
  id: string;
  uid: string | null;
  url: string | null;
  lang: 'en-gb';
  type: DocumentType;
  linked_documents: any[];
  slugs: string[];
  tags: string[];
  data: DataType;
}
