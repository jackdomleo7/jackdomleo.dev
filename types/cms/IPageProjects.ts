import { IImage } from "./IImage"
import { ILink } from "./ILink"
import { TRichText } from "./TRichText"

export interface IProject {
  name: string;
  url: ILink;
  project_type: 'standard' | 'mini' | 'website';
  short_description: TRichText;
  long_description: TRichText;
  tech_1: string | null;
  tech_2: string | null;
  tech_3: string | null;
  tech_4: string | null;
  thumbnail: IImage;
  mockup: IImage;
}

export interface IPageProjects {
  projects_intro: TRichText;
  mini_intro: TRichText;
  websites_intro: TRichText;
  projects: IProject[];
}