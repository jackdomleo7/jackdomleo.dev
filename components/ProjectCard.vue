<template>
  <component :is="project.url.url ? 'a' : 'div'" :href="project.url.url" :target="project.url.url && '_blank'" class="project-card">
    <nuxt-img provider="prismic" :src="project.thumbnail.url" :alt="project.thumbnail.alt" :width="project.thumbnail.dimensions.width" :height="project.thumbnail.dimensions.height" loading="lazy" class="project-card__img" />
    <div class="project-card__info">
      <h3 class="project-card__heading">
        {{ project.name }}
        <small v-if="project.project_type === 'website' && project.url.url">({{ project.url.url.replace('https://', '') }})</small>
      </h3>
      <ul v-if="project.tech_1 || project.tech_2 || project.tech_3 || project.tech_4" class="project-card__tags">
        <li v-if="project.tech_1" class="tag">{{ project.tech_1 }}</li>
        <li v-if="project.tech_2" class="tag">{{ project.tech_2 }}</li>
        <li v-if="project.tech_3" class="tag">{{ project.tech_3 }}</li>
        <li v-if="project.tech_4" class="tag">{{ project.tech_4 }}</li>
      </ul>
      <prismic-rich-text :field="project.short_description" class="project-card__description" />
    </div>
  </component>
</template>

<script lang="ts">
import Vue, { PropType } from 'vue'
import { IProject } from '@/types/cms/IPageProjects'

export default Vue.extend({
  name: 'ProjectCard',
  props: {
    project: {
      type: Object as PropType<IProject>,
      required: true
    }
  }
})
</script>

<style lang="scss" scoped>
.project-card {
  display: inline-flex;
  flex-direction: column;
  background-color: var(--colour-foreground-1);
  box-shadow: var(--shadow-standard);
  border-radius: 1.25rem;
  padding: 0.5rem;
  text-decoration: none;
  height: 100%;
  max-width: 31.25rem;

  &__img {
    border-radius: 0.75rem;
    width: 100%;
  }

  &__info {
    padding: 0.5rem;
  }

  &__heading {
    text-transform: uppercase;
    color: var(--colour-text-primary);
    letter-spacing: 0.02rem;
    font-size: var(--text-large);
    font-weight: 400;
    margin-top: 0.25rem;
    margin-bottom: 0.75rem;

    small {
      text-transform: lowercase;
      font-size: var(--text-small);
    }
  }

  &__tags {
    list-style-type: none;
    padding-left: 0;
    margin: 0;
  }

  &__description {
    margin-top: 1rem;
    font-size: var(--text-small);
    color: var(--colour-text-secondary);
    display: flex;
    flex-direction: column;
    gap: 0.5rem;

    ::v-deep p {
      margin: 0;
    }
  }
}
</style>
