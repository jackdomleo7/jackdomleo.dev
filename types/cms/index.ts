import { IPage } from "./IPage"
import { IPageHome } from "./IPageHome"
import { IPageAbout } from "./IPageAbout"
import { IPageProjects } from "./IPageProjects"
import { IPageBlog } from "./IPageBlog"

export {
  IPage,
  IPageHome,
  IPageAbout,
  IPageProjects,
  IPageBlog
}
