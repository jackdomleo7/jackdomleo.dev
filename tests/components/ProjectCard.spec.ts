import { mount, shallowMount } from '@vue/test-utils'
import { configureAxe, toHaveNoViolations } from 'jest-axe'
import ProjectCard from '@/components/ProjectCard.vue'
import { IProject } from '@/types/cms/IPageProjects'

const projects: Record<'checka11y'|'mywebsite'|'nourl', IProject> = {
  checka11y: {
    name: 'Checka11y.css',
    short_description: [
      {
        spans: [],
        text: 'A CSS stylesheet to quickly highlight a11y concerns',
        type: 'paragraph'
      }
    ],
    project_type: 'standard',
    url: {
      link_type: 'Web',
      url: 'https://checka11y.jackdomleo.dev'
    },
    thumbnail: {
      url: 'https://example.com/thumbnail.png',
      alt: 'Checka11y.css logo',
      copyright: null,
      dimensions: {
        height: 242,
        width: 484
      }
    },
    tech_1: 'SCSS',
    tech_2: 'Cypress',
    tech_3: null,
    tech_4: null,
    long_description: null,
    mockup: {
      url: null,
      alt: null,
      copyright: null,
      dimensions: {
        height: null,
        width: null
      }
    }
  },
  mywebsite: {
    name: 'My Website',
    short_description: [
      {
        spans: [],
        text: 'A fully responsive static website for my portfolio & blog.',
        type: 'paragraph'
      }
    ],
    project_type: 'website',
    url: {
      link_type: 'Web',
      url: 'https://jackdomleo.dev'
    },
    thumbnail: {
      url: 'https://example.com/thumbnail.png',
      alt: 'jackdomleo.dev thumbnail',
      copyright: null,
      dimensions: {
        height: 242,
        width: 484
      }
    },
    tech_1: 'Nuxt',
    tech_2: 'TypeScript',
    tech_3: 'SCSS',
    tech_4: null,
    long_description: [
      {
        spans: [],
        text: 'Lorem ipsum',
        type: 'paragraph'
      }
    ],
    mockup: {
      url: 'https://example.com/mockup.png',
      alt: 'jackdomleo.dev mockup',
      copyright: null,
      dimensions: {
        height: 374,
        width: 768
      }
    }
  },
  nourl: {
    name: 'Some project with no url',
    short_description: [
      {
        spans: [],
        text: 'A website with no url provided.',
        type: 'paragraph'
      }
    ],
    project_type: 'website',
    url: {
      link_type: 'Web',
      url: null
    },
    thumbnail: {
      url: 'https://example.com/thumbnail.png',
      alt: 'Some thumbnail',
      copyright: null,
      dimensions: {
        height: 242,
        width: 484
      }
    },
    tech_1: 'Nuxt',
    tech_2: 'TypeScript',
    tech_3: 'SCSS',
    tech_4: null,
    long_description: [
      {
        spans: [],
        text: 'Lorem ipsum',
        type: 'paragraph'
      }
    ],
    mockup: {
      url: 'https://example.com/mockup.png',
      alt: 'jackdomleo.dev mockup',
      copyright: null,
      dimensions: {
        height: 374,
        width: 768
      }
    }
  }
}

expect.extend(toHaveNoViolations)

const axe = configureAxe({
  rules: {
    region: { enabled: false } // Don't need to check because components are loaded in isolation
  }
})

describe('ProjectCard', () => {
  /**
   * Vue instance
   */

  it('is a Vue instance', () => {
    const wrapper = mount(ProjectCard, {
      propsData: {
        project: projects.mywebsite,
        headingLevel: 'h3'
      },
      stubs: ['prismic-rich-text', 'nuxt-img']
    })

    expect(wrapper.vm).toBeTruthy()
  })

  /**
   * Snapshots
   */

  it('matches snapshot when mounted', () => {
    const wrapper = mount(ProjectCard, {
      propsData: {
        project: projects.mywebsite,
        headingLevel: 'h3'
      },
      stubs: ['prismic-rich-text', 'nuxt-img']
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('matches snapshot when shallow mounted', () => {
    const wrapper = shallowMount(ProjectCard, {
      propsData: {
        project: projects.mywebsite,
        headingLevel: 'h3'
      },
      stubs: ['prismic-rich-text', 'nuxt-img']
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  /**
   * A11y
   */

  it('to have no obvious accessibility violations', async () => {
    const wrapper = mount(ProjectCard, {
      propsData: {
        project: projects.mywebsite,
        headingLevel: 'h3'
      },
      stubs: ['prismic-rich-text', 'nuxt-img']
    })

    await expect(axe(wrapper.element)).resolves.toHaveNoViolations()
  })

  /**
   * Props
   */

  it('renders an <a> with a href if a url exists', () => {
    const wrapper = shallowMount(ProjectCard, {
      propsData: {
        project: projects.mywebsite,
        headingLevel: 'h3'
      },
      stubs: ['prismic-rich-text', 'nuxt-img']
    })

    expect(wrapper.find('a.project-card').exists()).toBe(true)
    expect(wrapper.find('a.project-card').attributes('href')).toBe(projects.mywebsite.url.url)
  })

  it('renders a <div> if a url is not present', () => {
    const wrapper = shallowMount(ProjectCard, {
      propsData: {
        project: projects.nourl,
        headingLevel: 'h3'
      },
      stubs: ['prismic-rich-text', 'nuxt-img']
    })

    expect(wrapper.find('div.project-card').exists()).toBe(true)
  })

  it('shows url if the project is a website', () => {
    const wrapper = shallowMount(ProjectCard, {
      propsData: {
        project: projects.mywebsite,
        headingLevel: 'h3'
      },
      stubs: ['prismic-rich-text', 'nuxt-img']
    })

    expect(wrapper.find('h3.project-card__heading small').exists()).toBe(true)
    expect(wrapper.find('h3.project-card__heading small').text()).toContain(projects.mywebsite.url.url.replace('https://', ''))
  })

  it('does not show url if the project is not a website', () => {
    const wrapper = shallowMount(ProjectCard, {
      propsData: {
        project: projects.checka11y,
        headingLevel: 'h3'
      },
      stubs: ['prismic-rich-text', 'nuxt-img']
    })

    expect(wrapper.find('h3.project-card__heading small').exists()).toBe(false)
  })
})
