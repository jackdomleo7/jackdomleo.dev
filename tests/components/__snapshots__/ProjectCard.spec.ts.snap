// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectCard matches snapshot when mounted 1`] = `
<a
  class="project-card"
  href="https://jackdomleo.dev"
  target="_blank"
>
  <nuxt-img-stub
    alt="jackdomleo.dev thumbnail"
    class="project-card__img"
    height="242"
    loading="lazy"
    provider="prismic"
    src="https://example.com/thumbnail.png"
    width="484"
  />
   
  <div
    class="project-card__info"
  >
    <h3
      class="project-card__heading"
    >
      
      My Website
      
      <small>
        (jackdomleo.dev)
      </small>
    </h3>
     
    <ul
      class="project-card__tags"
    >
      <li
        class="tag"
      >
        Nuxt
      </li>
       
      <li
        class="tag"
      >
        TypeScript
      </li>
       
      <li
        class="tag"
      >
        SCSS
      </li>
       
      <!---->
    </ul>
     
    <prismic-rich-text-stub
      class="project-card__description"
      field="[object Object]"
    />
  </div>
</a>
`;

exports[`ProjectCard matches snapshot when shallow mounted 1`] = `
<a
  class="project-card"
  href="https://jackdomleo.dev"
  target="_blank"
>
  <nuxt-img-stub
    alt="jackdomleo.dev thumbnail"
    class="project-card__img"
    height="242"
    loading="lazy"
    provider="prismic"
    src="https://example.com/thumbnail.png"
    width="484"
  />
   
  <div
    class="project-card__info"
  >
    <h3
      class="project-card__heading"
    >
      
      My Website
      
      <small>
        (jackdomleo.dev)
      </small>
    </h3>
     
    <ul
      class="project-card__tags"
    >
      <li
        class="tag"
      >
        Nuxt
      </li>
       
      <li
        class="tag"
      >
        TypeScript
      </li>
       
      <li
        class="tag"
      >
        SCSS
      </li>
       
      <!---->
    </ul>
     
    <prismic-rich-text-stub
      class="project-card__description"
      field="[object Object]"
    />
  </div>
</a>
`;
